### PR TITLE
TA offline signing

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

File: ta/arch/arm/link.mk [hunk 1]
```diff
@@ -103,12 +103,21 @@ ifeq ($(CFG_ENCRYPT_TA),y)
 crypt-args$(user-ta-uuid) := --enc-key $(TA_ENC_KEY)
 cmd-echo$(user-ta-uuid) := SIGNENC
 endif
+ifeq ($(TA_SIGN_METHOD),stitch)
+# Offline signing
+TA_SIG_DIR ?= signatures
+crypt-args$(user-ta-uuid) := --sig $(TA_SIG_DIR)/$(user-ta-uuid).sig
+sign-method$(user-ta-uuid) := stitch
+cmd-echo$(user-ta-uuid) := SIGNST #
```
Comment (wamserma) [inline ta/arch/arm/link.mk:111]: stray `#`

**Comment by jforissier** [comment]
> @jforissier Will you also add a variable for the signing scheme or I should have it locally? More info [here](https://github.com/OP-TEE/optee_os/pull/4614#discussion_r635880695).

Missed that sorry. Patch updated.

**Comment by aleksandarn-ee** [comment]
> > @jforissier Will you also add a variable for the signing scheme or I should have it locally? More info [here](https://github.com/OP-TEE/optee_os/pull/4614#discussion_r635880695).
> 
> Missed that sorry. Patch updated.

No worries.

I tested the patch, the building process seems great now, however one thing is driving me crazy and I have no idea what is going on. After building the OPTEE core and TA elf files, I am stitching the signatures with the following command (we use the default `TA_SIG_DIR` directory):

```
TA_SIGN_METHOD=stitch TA_SIG_ALGO=TEE_ALG_RSASSA_PKCS1_V1_5_SHA256 make ta
```

However, the error is as following:

```
alenik@nb152:~/projects/github/optee/optee_os_jforissier$ TA_SIGN_METHOD=stitch TA_SIG_ALGO=TEE_ALG_RSASSA_PKCS1_V1_5_SHA256 make ta
  CHK     out/arm-plat-vexpress/conf.mk
  CHK     out/arm-plat-vexpress/include/generated/conf.h
  SIGNST  out/arm-plat-vexpress/ta/avb/023f8f1a-292a-432b-8fc4-de8471358067.ta
ERROR:sign_encrypt.py:No signature file found. Please sign
 023f8f1a-292a-432b-8fc4-de8471358067.dig
offline and place the signature at 
 /023f8f1a-292a-432b-8fc4-de8471358067.sig
or pass a different location using the --sig argument.

make: *** [ta/arch/arm/link.mk:139: out/arm-plat-vexpress/ta/avb/023f8f1a-292a-432b-8fc4-de8471358067.ta] Fehler 1
```

From the output above, the variable `TA_SIG_DIR` seems to not be even set (notice `signatures` is missing here ` /023f8f1a-292a-432b-8fc4-de8471358067.sig`). 

I debugged a little further and apparently all the variables in this if statement are **not** set. Interestingly enough when I compile with `make -j4` `TA_SIG_DIR` is set to `signatures` but only for the `trusted_keys` and `pkcs11` TAs, which are built after the `avb` TA. Finally, compiling the TAs with `TA_SIG_DIR=./signatures` works.

Am I missing something here? I know the problem is not in the OPTEE Makefile, since `ifeq` is used throughout the code and people would notice the problem, but rather how I call the `make` command.

**Comment by jforissier** [comment]
Updated, @aleksandarn-ee please try again :)

File: core/arch/arm/kernel/link.mk [hunk 2]
```diff
@@ -216,6 +216,9 @@ $(link-out-dir)/tee.symb_sizes: $(link-out-dir)/tee.elf
 	@$(cmd-echo-silent) '  GEN     $@'
 	$(q)$(NMcore) --print-size --reverse-sort --size-sort $< > $@
 
+tee_bin: $(link-out-dir)/tee-header_v2.bin $(link-out-dir)/tee-pager_v2.bin \
+	 $(link-out-dir)/tee-pageable_v2.bin
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/kernel/link.mk:220]: +1   platform can also boot from `tee.bin`. I agree it should be built.

File: ta/arch/arm/link.mk [hunk 3]
```diff
@@ -73,6 +73,8 @@ link-script-cppflags-$(sm) := \
 
 link-script-pp-makefiles$(sm) = $(filter-out %.d %.cmd,$(MAKEFILE_LIST))
 
+TA_SIG_DIR ?= signatures
```
Comment (etienne-lms) [inline ta/arch/arm/link.mk:76]: Should be documented, somehting like:
`Path of the directory that stores file <uuid>.sig files each storing the signature to be stitched to <uuid>.elf file into a loadable signed <uuid>.ta image.`

**Review by etienne-lms — COMMENTED** [review]
"Add 'ta_elf' target"
"Add 'ta' target"
"Add 'tee_bin' target" (with @aleksandarn-ee's comment addressed)
Maybe prefix each header line with: `make: ` or `build: `

**(edited):** Or change header line to "Add 'xxx' make target".

File: ta/arch/arm/link.mk [hunk 4]
```diff
@@ -103,12 +105,23 @@ ifeq ($(CFG_ENCRYPT_TA),y)
 crypt-args$(user-ta-uuid) := --enc-key $(TA_ENC_KEY)
 cmd-echo$(user-ta-uuid) := SIGNENC
 endif
+ifeq ($(TA_SIGN_METHOD),stitch)
```
Comment (etienne-lms) [inline ta/arch/arm/link.mk:108]: Should we set a default value to TA_SIGN_METHOD, in mk/config.mk (i.e. `TA_SIGN_METHOD=sign-enc`)?

Comment (wamserma) [reply ta/arch/arm/link.mk:108]: Not necessary. The script defaults to signing.

File: ta/arch/arm/link.mk [hunk 5]
```diff
@@ -103,12 +105,23 @@ ifeq ($(CFG_ENCRYPT_TA),y)
...
+# Offline signing
+crypt-args$(user-ta-uuid) := --sig $(TA_SIG_DIR)/$(user-ta-uuid).sig
```
[the remark this answers is not in the document]
Comment (jforissier) [reply ta/arch/arm/link.mk:110]: Good catch, will fix.

File: core/arch/arm/kernel/link.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (jforissier) [reply core/arch/arm/kernel/link.mk:220]: OK

File: ta/arch/arm/link.mk (hunk 3 above)
Comment (jforissier) [reply ta/arch/arm/link.mk:76]: I will add a comment. Yes, all paths are relative to the top level directory of the project.

File: ta/arch/arm/link.mk (hunk 4 above)
Comment (jforissier) [reply ta/arch/arm/link.mk:108]: `sign_enc.py` assumes sign-enc is the default command, so I'd rather keep it as is for simplicity.

**Comment by jforissier** [comment]
Update

File: ta/arch/arm/link.mk (hunk 4 above)
Comment (etienne-lms) [reply ta/arch/arm/link.mk:110]: It makes the makefiles implementation not that obvious i think. TA_SIGN_METHOD referred only at 2 places, only 1 value found, alternative is its default non-existence empty value.
But i'm fine as-is if you think it's easier to maintain that way.

**Comment by jforissier** [comment]
@jbech-linaro is IBART missing `sudo apt install libgmp-dev`? First time I see this error... :confused:

**Comment by jbech-linaro** [comment]
> @jbech-linaro is IBART missing `sudo apt install libgmp-dev`? First time I see this error...

Yes, weird, but I've installed the package now. Thanks!

File: ta/arch/arm/link.mk (hunk 3 above)
Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk:78]: > Path of the directory that stores file <uuid>.sig files each storing the signature to be stitched to <uuid>.elf file into a loadable signed <uuid>.ta image.

Aren't signatures added to the `<uuid>.stripped.elf` files?

**Comment by aleksandarn-ee** [comment]
@jforissier Does the OPTEE documentation needs to be updated?

**Comment by wamserma** [comment]
> @jforissier Does the OPTEE documentation needs to be updated?

Yes, here https://optee.readthedocs.io/en/latest/building/trusted_applications.html#offline-signing-of-tas
I can take care of the doc once this PR is merged.

**Comment by jforissier** [comment]
@wamserma that would be great, thanks

**Comment by vesajaaskelainen** [comment]
Just a comment on what we do.

We use devel keys during building process for signing and then re-sign the elf image in our code signing environment. This way we know exactly what we signed and have a trace. These hash signing things do not leave the trace.

**Comment by aleksandarn-ee** [comment]
> Just a comment on what we do.
> 
> We use devel keys during building process for signing and then re-sign the elf image in our code signing environment. This way we know exactly what we signed and have a trace. These hash signing things do not leave the trace.

@vesajaaskelainen Doesn't the devel public key get embedded into the OPTEE core image during building in this case?

**Comment by vesajaaskelainen** [comment]
> > Just a comment on what we do.
> > We use devel keys during building process for signing and then re-sign the elf image in our code signing environment. This way we know exactly what we signed and have a trace. These hash signing things do not leave the trace.
> 
> @vesajaaskelainen Doesn't the devel public key get embedded into the OPTEE core image during building in this case?

Nope:
https://github.com/OP-TEE/optee_os/commit/3ecd96bbf821d170671cc19a9bb6dc3eba5b9136 - core: Add support to import external TA signing public key

So in devel builds for TEE OS the keys will be the same -- but when preparing for TEE OS release we inject real TA release signing public key in TEE OS -- then we can first test TA with devel TEE OS and once verified to be OK then we can re-sign the TA binary with release keys and then use it with TEE OS with TA release public key.

Only catch with this is that both TA signing keys must be of same size so that signature takes same amount of data. Eg. you cannot mix 2k and 4k keys.

**Comment by aleksandarn-ee** [comment]
> Nope:
3ecd96b - core: Add support to import external TA signing public key

Didn't see this commit until now, it makes a lot of sense to go this way.

> Only catch with this is that both TA signing keys must be of same size so that signature takes same amount of data. Eg. you cannot mix 2k and 4k keys.

I dont understand this fully. Why is it a problem if the devel keys are e.g. RSA2048 and the release keys are RSA4096? The public key that goes into the OPTEE OS will be an RSA4096 public key and the TAs will be signed with the RSA4096 private key? Unless the OPTEE build fails if the release public key (TA_PUBLIC_KEY) and devel private key (TA_SIGN_KEY) do not match in size? 

@jforissier Considering the commit 3ecd96b, do we need this PR? Maybe only the new Make Targets? Either way I would update the OPTEE documentation so it nicely describes the commit 3ecd96b.

**Comment by jforissier** [comment]
@aleksandarn-ee 

> @jforissier Considering the commit [3ecd96b](https://github.com/OP-TEE/optee_os/commit/3ecd96bbf821d170671cc19a9bb6dc3eba5b9136), do we need this PR? Maybe only the new Make Targets? Either way I would update the OPTEE documentation so it nicely describes the commit [3ecd96b](https://github.com/OP-TEE/optee_os/commit/3ecd96bbf821d170671cc19a9bb6dc3eba5b9136).

I suggest you propose a doc update and we'll see if any of the commits in this PR bring any value. @vesajaaskelainen thanks for chiming in!

**Comment by aleksandarn-ee** [comment]
> I suggest you propose a doc update and we'll see if any of the commits in this PR bring any value. 

@jforissier  @wamserma Something like [this](https://github.com/OP-TEE/optee_docs/pull/125)? Feel free to comment and update by yourself.

**Comment by jforissier** [comment]
> [...] we'll see if any of the commits in this PR bring any value [...]

Apparently not, let's close this.

**Pull request closed without merging**